Allow for other prefixes

whot opened the pull request:
cc @bentiss

whot committed:
Allow for other prefixes in the form of "X:" where X is anything

This is prep work for hidutils/hid-recorder#25 where we'll add the "B:"
prefix for BPF data. In the future we may add other prefixes too so
let's just ignore anything that looks correct but is unknown.

whot committed:
Change parse() to take a string iterator so we can test it easier

Let parse() take a string iterator instead of a file so we can easier pass
test values into it.

whot committed:
Handle various missing bits from the recording a bit more gracefully

bentiss commented on src/main.rs:
```diff
@@ -101,6 +102,20 @@ fn parse(lines: &[String], mut stderr: impl std::io::Write) -> Result<Recording>
                 });
             }
+            Some((prefix, _)) => {
+                if prefix.len() == 2 && prefix.chars().nth(1).unwrap_or(' ') == ':' {
```
```
if prefix.len() == 2 && prefix.ends_with(":") {
```

?

You could also split at `: ` and then match for the first char instead...

bentiss commented on src/main.rs:
```diff
@@ -62,18 +62,12 @@ fn data_decode(str: &str) -> Result<(usize, Vec<u8>)> {
-fn parse(path: &Path) -> Result<Recording> {
-    let f = std::fs::File::open(path)?;
-    let lines = std::io::BufReader::new(f).lines();
+fn parse(lines: &[String], mut stderr: impl std::io::Write) -> Result<Recording> {
```
It was already the case previously, but you likely want to lazy parse the incoming file. If the file is too big because it was recorded for too long you should just parse the lines as they com in, not load the entire file in memory first.

So IMO your `parse()` should take a single line and return a boolean to say continue/stop, and you should call this in a `.map()` call.

whot replied:
done now, took a while until I grasped it but the key was to use `impl Iterator` as function argument

bentiss reviewed (approved):
few nitpicks but otherwise seems to be good (though I'm not that familiar with the existing code)

whot committed:
Split parser by line

This allows more precise and focused tests.

whot committed:
Switch to a builder pattern for the Recording

whot committed:
Rename a function to be more precise

And add some tests while we're here.

whot committed:
Print the recording's line number in the error message

This requires that we move the filter into parse(), otherwise we're not
counting comment or empty lines.
